### ISSUE TITLE
LEDENET UFO

### ISSUE THREAD

**Issue by ckreed** [steve228uk/homebridge-magichome#6]
I can cycle power on the strip, but I cannot control brightness or color(s).  This uses the MagicHome app just fine, but homebridge only controls the power on/off.

Here is what I have:
[https://www.amazon.com/LEDENET-Controller-Android-Smartphone-Control/dp/B00MDKOSN0/ref=redir_mobile_desktop?_encoding=UTF8&keywords=ledenet%20ufo&pi=SY200_QL40&qid=1482177267&ref_=mp_s_a_1_1&sr=8-1](url)

**Comment by steve228uk** [comment]
Thanks @ckreed. Guessing you've tried the various configs? I'm gonna buy one of these next week and see if I can figure it out.

**Comment by ckreed** [comment]
Yes, I tried RGB, RGBW (which is what this is supposed to be) and RGBWW. I also tried those in combination with the purewhite setting.   After posting this I found a ruby gem ledenet-ufo that I can run via command line that works perfectly. But I lack the skill in integrating something like this as a homebridge plugin.  It might help you identify the differences.  Thanks much!!   

[https://github.com/sidoh/ledenet_api ](url)

**Comment by steve228uk** [comment]
@ckreed That link doesn't seem to be working. If you can send it over I'll have a look to see what's different.

**Comment by ckreed** [comment]
So strange - clicking the link results in a 404, copy/paste into browser will take you there.  I am obviously comment challenged some how (I cannot get the link to paste properly here).  Just put that address in your browser and it should work.  Sorry.

**Comment by ghost** [comment]
Hey first of all thanks for the great module!
I also have the Magic Home Ufo. The on/off function is working great, but I have the same problem like steve. Is there already a solution?
Thanks!

**Comment by steve228uk** [comment]
@ckreed @HangJezz Okay, this is the file that's doing the work:

https://github.com/sidoh/ledenet_api/blob/b6f0c5cba03c535cdc877ac2f9787e88763d16be/lib/ledenet/packets/update_color_request.rb#L10-L20

This is what my plugin is currently doing

https://github.com/steve228uk/homebridge-magichome/blob/master/flux_led.py#L576-L590

It looks like RGBWW is the correct pattern to use on this. 

Are you able to try the `flux_led.py` script from this plugin via the command line and see if setting that as RGBWW works?

**Comment by steve228uk** [comment]
It looks like @tmarcus02 has created a fork with multiple RGBWW options.

https://github.com/tmarcus02/homebridge-magichome/commit/3240b5e534796a4a9b471705b31084b4022d611b

Perhaps they can weigh in as to whether they got the UFO to work? I'd be more than happy to merge a PR if so 😊

**Comment by ghost** [comment]
I will try that tomorrow.. Thanks for your fast answer! I am sure to get the UFO working xD

**Comment by ckreed** [comment]
I tried the flux_led.py before posting here.  It can discover and turn on and off, but I cannot control the color or brightness of the warm white.  One thing I noticed is that if I use the constant Red or "#FF0000" with the -c parameter I get an error of bad color specification.  Using 255,0,0 notation is accepted without an error but nothing happens.   The Ruby Gem that I mentioned earlier works perfectly.  Hope that helps!

**Comment by ckreed** [comment]
Here is a screen shot of the command line.

![screen shot 2017-02-01 at 1 54 08 pm](https://cloud.githubusercontent.com/assets/25210881/22523745/08a9e5f2-e886-11e6-9569-9f95beb82ddb.png)

and the errors on the color

![screen shot 2017-02-01 at 1 56 10 pm](https://cloud.githubusercontent.com/assets/25210881/22523797/43913a9e-e886-11e6-8ef9-6c71b235138f.png)

**Comment by tmarcus02** [comment]
The fork I created was to add in support for the following controllers which I am using.

For the controller [XCSOURCE® Magic UFO WiFi LED Controller](https://www.amazon.co.uk/gp/product/B00Q7STR4E/ref=oh_aui_detailpage_o07_s00?ie=UTF8&psc=1) specify RGBWWv3 in the homebridge config file i.e.   "setup": "RGBWWv3"

For the controller [Sunix® Wireless WiFi RGB/RGBWWCW LED Strip Light Controller](https://www.amazon.co.uk/gp/product/B01J5A7ABW/ref=oh_aui_detailpage_o08_s00?ie=UTF8&psc=1) specify RGBWWv1 in the homebridge config file i.e.   "setup": "RGBWWv1"

**Comment by tmarcus02** [comment]
If these options don't work for your controller you need to do a packet capture when using the Magic Home app to change the colour. This will help identify how the data in the packet being sent to the controller is formatted.

If you have a iPhone and a Mac, you can follow these instructions:-

1. Connect your iOS device to your Mac via USB.

2. Get the UDID for the connected device from iTunes or organiser.

3. Open terminal in your Mac

4. type the following commands in the terminal

`$ ifconfig -l `// First get the current list of interfaces.

`$ rvictl -s <udid>` // Then run the tool with the UDID of the device.

 // This adds a new virtual network interface rvi0.

`$ ifconfig -l  `// Get the list of interfaces again, and you can see the new virtual network interface, rvi0, added by the previous command.

5. Open Wireshark and run a capture on the rvi0 interface. Should look like this on the capture:-

![wireshark2](https://cloud.githubusercontent.com/assets/23484668/22535145/f53e05f0-e8f0-11e6-8fe5-52359a0844e0.jpg)

6. To remove the virtual interface, type the following command in the terminal

`$ rvictl -x  <udid>`

**Comment by ghost** [comment]
Yes I also have the XCSOURCE UFO. Are there differences between this Ufos? There are a lot of Ufos on the market. I tought they all have the same hard- and software. 🙈

**Comment by steve228uk** [comment]
@tmarcus02 Thanks for confirming. I have that second device but haven't tried RGBWW with it yet. Are you cool with submitting a PR for your changes? If not I'm happy to add them manually.

**Comment by tmarcus02** [comment]
@HangJezz - I don't know if they are all the same hardware and software, when using the Magic Home App it identifies the type of the controller and the version, So I don't know if there is a Magic UFO available which are other versions - mine is v3

![img_0063](https://cloud.githubusercontent.com/assets/23484668/22547512/e0a818bc-e939-11e6-8fc5-3586688a5437.jpg)

**Comment by tmarcus02** [comment]
@steve228uk - I haven't tested yet using a RGBWW strip, I am just running RGB on both of my controllers. I test it out to confirm there is no further changes required for these type of controllers

**Comment by ckreed** [comment]
Well crap.  Even with @tmarcus02 version (and the RGBWWv3 setup) mine still doesn't work.  I have setup wireshark and can see the data. Here is what I got when setting the color to R=131,G=92,B=87.  Not sure if this helps.

![screen shot 2017-02-02 at 6 40 09 am](https://cloud.githubusercontent.com/assets/25210881/22549931/630218f2-e913-11e6-8a59-5af6ece4430f.png)

**Comment by tmarcus02** [comment]
@ckreed - Looking at the data 3183515100000f6531835c5700000f76, the Magic App will send larger packets when you slide you finger around the colour wheel, A single colour change the data is 8 byte in length.

So going by the RGB colour you selected the data is made up like this 31835c5700000f76

31 - `persist=True`
83 - R= 131
5c - G= 92
57 - B = 87

00000f - matches the RGBWW padding, so setting `"setup": "RGBWW"` in the home bridge config file should work for your controller.

**Comment by ckreed** [comment]
@tmarcus02  MagicHome app shows v3.  I tried RGBWW, RGBWWv2, RGBWWv3.  *shrug*  Not sure what is up.  Still only turns on and off. No brightness or colors.  The Ruby Gem ledent-ufo version (command line) works fine.  The flux_led.py version (command line) does not.   I appreciate everyone's help in looking at this.  I guess it will be easier to just go buy one that has been tested. 
:-(

**Comment by ghost** [comment]
So I have 3 UFOS at home and all have the version v3. So now I have to edit the flux_led.py and the config to RGBWWv3? What are the Commands for testing the flux thing in command line?

**Comment by ckreed** [comment]
@HangJezz 
![screen shot 2017-02-02 at 9 15 46 am](https://cloud.githubusercontent.com/assets/25210881/22555140/4f6e9c06-e928-11e6-8268-74b3af7091a6.png)

**Comment by ghost** [comment]
ahh ok thanks!

**Comment by ghost** [comment]
Ok now I have the same problem. I can set the stripe on and off, set the color, only warm white is not working. I am using the modified script and RGBWWv3

**Comment by steve228uk** [comment]
@ckreed Does the colour wheel show at all in home.app?

**Comment by ghost** [comment]
Every time I send a color, the color is working, but warm white turns always to 100%. Funny Bug : /

**Comment by tmarcus02** [comment]
@ckreed - When using the command line to execute the Flux_led.py script, are you stating the options like the following, I have tested the Flux_led.py this way and it is working ok

![untitled](https://cloud.githubusercontent.com/assets/23484668/22557759/6cf4c4f4-e963-11e6-9e75-1890ae5c9b83.jpg)

**Comment by ckreed** [comment]
@steve228uk Yes, I can pull up the brightness slider and the color wheel (when editing the pre-selects), but I can only turn on/off the strip.  No color changes, no warm white changes, no brightness changes on either (RGB or WW).  Behaves just like the flux_led.py script.  Another note, upon startup of homebridge, it looks like it goes out and gets the status of the strip.  It figures out that the strip is on/off, but it always returns (255,100,50) as the current color, which is not correct.

**Comment by ckreed** [comment]
![screen shot 2017-02-02 at 10 31 01 am](https://cloud.githubusercontent.com/assets/25210881/22558277/bcb35360-e932-11e6-84ef-eb2f875791ec.png)
@tmarcus02 Yes, that is how I have tried the command.  Here is a screenshot issuing a color change.  It executes with no errors but nothing happens.  Then I immediately check the color and you can see it never set the color.  So the -info works, but making any change (other than on/off) does not work.

![screen shot 2017-02-02 at 10 26 06 am](https://cloud.githubusercontent.com/assets/25210881/22558144/4d62c1e4-e932-11e6-90b7-0f7dddfdbc4a.png)

**Comment by tmarcus02** [comment]
@HangJezz - Glad to see the colour change is now working, I have not been using RGBWW just RGB so the modification I made was to just get the colour change working quickly via homekit for my controllers

I have today connected a white strip up to the Magic UFO controller and I'm getting the same results you are getting. So I'm looking into it and will update.

**Comment by ghost** [comment]
@tmarcus02 Thank you very much! If there is anything I can doo to help just say it!

**Comment by tmarcus02** [comment]
Apologies @ckreed my mistake your padding is different 00000f - RGBWW is 0000f0.

I'll modify the code and add in the option RGBWWv3a

**Comment by tmarcus02** [comment]
Ok updated Flux_led.py, try it again and set the setup option to RGBWWv3a

**Comment by ghost** [comment]
Where can I get it?

**Comment by ghost** [comment]
Ok I found it... Sry I am new on github xD

**Comment by ghost** [comment]
So the color is working great now, but the brightness of warm white does not work or is this my mistake?
This is my Command:
python flux_led.py 192.168.178.49 -x RGBWWv3a -w 20

**Comment by ckreed** [comment]
@tmarcus02 @HangJezz  Confirmed  flux_led.py color now works also on/off!  Warm white does not work with flux_led.py. Also the home.app is completely broke. Cannot turn on/off, cannot color change, cannot warm white change.

**Comment by tmarcus02** [comment]
Sry @HangJezz I was talking about the problem @ckreed was having with his controller which seems to be different and wasn't able to change the colour yet, So I added the option RGBWWv3a for his controller.

I'm still thinking about the Warm white option and how to handle it, as it is setting white to 100% when setting the colour.

The way I am think about it is if you were selecting white as the colour you would want the white LEDs to illuminate not at 100% but at the same brightness level as the RGB LEDs. Also when selecting a colour you would not want the white LEDs to illuminate.

@steve228uk can you confirm what way the other controllers work via homebridge-magichome when using RGBWW. I have attached screenshots of the homekit controls for magichome plugin but there doesn't seem to be a control to adjust the warm white brightness in the same way there is in the Magic Home app

![img_0064](https://cloud.githubusercontent.com/assets/23484668/22562028/bed012d8-e972-11e6-933b-742e671d76e3.jpg)
![img_0066](https://cloud.githubusercontent.com/assets/23484668/22562030/c093ea86-e972-11e6-88f3-a41db4788a59.jpg)
![img_0067](https://cloud.githubusercontent.com/assets/23484668/22562032/c32aca6c-e972-11e6-8d6f-566deb74223f.jpg)

![img_0068](https://cloud.githubusercontent.com/assets/23484668/22562263/b45c82ae-e973-11e6-9d43-f192afca6fe9.jpg)

**Comment by tmarcus02** [comment]
@ckreed You may need to remove the old homebridge-magichome plugin if it is not working via homekit

Step 1 `sudo npm uninstall -g homebridge-magichome`

Step 2 `sudo npm install -g tmarcus02/homebridge-magichome`

Once the warm white bugs are sorted I'll submit a PR so @steve228uk can approve the changes and then the install directory would return to homebridge-magichome using the command `sudo npm install -g homebridge-magichome`

**Comment by ghost** [comment]
I am very happy, that the colors and the brightness are working now. I can control my LED-Strips with Siri now xD. I would say the only option with homekit is to use the the normal white, but I am not a expert... Finallay thanks for the fast answers and the good work!

**Comment by ckreed** [comment]
@tmarcus02 That did it!  Very nice.  Appreciate it.  I would think that there is some homekit interfaces that now support the warm white.  Probably just have to figure out how to publish that feature to homekit for discovery.  Thank you very much for all the effort.  I even learned a few things.  :-)

**Comment by steve228uk** [comment]
@tmarcus02 Currently the light temperature isn't controllable at all. The temperature dial returns a mix of HSL just like the colour dial and that's what's converted into temperature by other bulbs such as Philips Hue.

You can see here that if the `purewhite` option is set to true and the saturation and hue are 0 (the middle of the colour wheel) then it will send the warm white option

https://github.com/steve228uk/homebridge-magichome/blob/master/index.js#L114-L117

We could potentially set a range based on what's sent from the temperature wheel to set either warm white or cool white based on the config and the temperature selected.

**Comment by ghost** [comment]
@tmarcus02 
Sadly your fix doesn't work for me. Can't controll anything via the app. If I try to run the script this message appears:
```
sudo python /usr/lib/node_modules/homebridge-magichome/flux_led.py 192.168.2.104 -x RGBWWv1 -w 20
  File "/usr/lib/node_modules/homebridge-magichome/flux_led.py", line 604
    def setRgb(self, r,g,b, persist=True, setup="RGBW"):
                                                       ^
IndentationError: unindent does not match any outer indentation level
```

I got this version:
https://www.amazon.de/gp/product/B01J5A5YYM/ref=oh_aui_detailpage_o00_s01?ie=UTF8&psc=1

**Comment by ckreed** [comment]
@whitechapplin Hmmmm, that is a python error.  You might try to download or reload again?  The script is not running due to spaces or tabs.  I did not see anything similar when I ran the script (though I am not a python expert).

**Comment by ckreed** [comment]
@steve228uk @tmarcus02 Just an FYI - one package I had to add to my python environment was webcolors (I used pip to install).  After I did that all complaints about colors not defined went away.  Not sure if that can be packaged in or if just a note should be added saying if you want to use "Red" or color names you must have the webcolors package installed.

**Comment by tmarcus02** [comment]
I made some changes last night around 00:00 to 01:30, these were to fix white control on RGBWW & RGBW led strips and I created a new branch to test them.

The Magic UFO now has the setup options of RGBW, RGBWv3 & RGBWv3a.

I tested the changes which were working, so merged the branch into the master but it seems to have messed things up as when I installed the new version on my server it loaded into home bridge ok but no features are working i.e. on/off, change colour, etc.. and I wasn't getting any errors on the homebridge console.

It isn't even getting the status of the Controllers via homebridge, I revert the changes but I'm still having the same problem, I'm update my Server currently so going to retest again this morning.

**Comment by ghost** [comment]
@tmarcus02 Seems like your newest fixes doesn't work :(

**Comment by tmarcus02** [comment]
Ok I'm still having problems but I think its to do with Homebridge. 

I have removed the fix for white control on RGBWW & RGBW strips, The changes were made on the scripts flux_led.py and index.js. I have reset any changes to index.js by replacing it with a unmodified version.

The setup options for the Magic UFO controller are still in place these are RGBW, RGBWv3 & RGBWv3a I have tested the flux_led.py script and it is working correctly.

I can't test via homekit currently, the latest is I can't add Homebridge to the Home.app even after removing the homebridge-magichome plugin.

**Comment by ghost** [comment]
@tmarcus02 The script is working correctly for me right now. But sadly there is still no response via homekit

**Comment by tmarcus02** [comment]
Updates @whitechapplin @ckreed @HangJezz 

Was able to get Homebridge working again with the Home app, Details were on the homebridge homepage to do this.

 > Two reasons why Homebridge may not be discoverable:
> 
> 1. Homebridge server thinks it's been paired with, but iOS thinks otherwise. Fix: deleted persist/ directory which is next to your config.json.
> 2.  iOS device has gotten your Homebridge username (looks like a MAC address) "stuck" somehow, where it's in the database but inactive. Fix: change your username in the "bridge" section of config.json to be some new 

I have been able to sort out the main issue with been unable to control (on/off, change colour, etc) from the Home app. For some reason I had to declare "python" to execute the flux_led.py from index.js which wasn't necessary before.

```
MagicHomeAccessory.prototype.sendCommand = function(command, callback)
{
	var exec = require('child_process').exec;
	var cmd = "python " +  __dirname + '/flux_led.py ' + this.ip + ' ' + command;

```

White control is work currently via the way @steve228uk detailed before and I think it would be better if the temperature control adjusted white between warm and cool and the brightness control could dim white LEDs in this case.

> You can see here that if the purewhite option is set to true and the saturation and hue are 0 (the middle of the colour wheel) then it will send the warm white option
> 
> https://github.com/steve228uk/homebridge-magichome/blob/master/index.js#L114-L117
> 
> We could potentially set a range based on what's sent from the temperature wheel to set either warm white or cool white based on the config and the temperature selected.

The controllers setup options are RGBW, RGBWv3, RGBWW & RGBWWv1

**Comment by ckreed** [comment]
@tmarcus02 So did we drop the v3a that was working for the UFO?  It no longer works as the previous version did with the RGBWWv3a setting.  

Also, I can confirm that the FLUX WiFI Smart LED Light does work with the RGBW setting.

**Comment by tmarcus02** [comment]
@ckreed Yes I cleaned up the options, RGBWv3 should work for your controller.

**Comment by ckreed** [comment]
@tmarcus02   :-(   Sorry, the Home app and flux_led.py will turn on/off the strip, but that is all.  No color or brightness control from either (when using RGBWv3, or any of the others).

**Comment by tmarcus02** [comment]
I checked back in the history and the code for RGBWWv3a is the same as RGBWv3.

```
if setup == "RGBWWv3a":
    msg.append(0x00)
    msg.append(0x00)
    msg.append(0x0f)
```

```
if setup == "RGBWv3":
    msg.append(0x00)
    msg.append(0x00)
    msg.append(0x0f)
```

The only thing I can think of is try uninstalling and re-installing using the following commands

Step 1    `sudo npm uninstall -g --unsafe-perm homebridge-magichome`

Step 2   `sudo npm install -g --unsafe-perm tmarcus02/homebridge-magichome `

**Comment by ckreed** [comment]
Perfect!!  That did the trick. Well done.

**Comment by Klangen82** [comment]
So I have read this thread an maybe I´m a bit of a noob but could someone please add a guide how you got the UFO to work? I have tested some configs etc. but don´t get a stable and working solution...

**Comment by Klangen82** [comment]
So this is how I hot it to work on a Magic UFO led.

1. I open the Magic Home app on my iphone, whent to the settings and then check the version I have. (in my case version v3)
![fullsizerender](https://cloud.githubusercontent.com/assets/25564168/24079726/277fee40-0c8f-11e7-9545-27be05fcb6be.jpg)

2. Installed HomeBridge MagicHome plugin via `sudo npm install -g --unsafe-perm tmarcus02/homebridge-magichome`

3. Added the MagicHome to the config and depending on your version you can have RGBW, RGBWv3 & RGBWv3a an an option.
![image](https://cloud.githubusercontent.com/assets/25564168/24079749/95e8d798-0c8f-11e7-9654-b1252fcd06c3.png)

If someone want to add this to the readme please be my guest:)

**Comment by steve228uk** [comment]
@Klangen82 I'll get this added and merge in @tmarcus02's changes today. Thanks for this.

**Comment by GatoPharaoh** [comment]
thanks @Klangen82 - works like a charm with @tmarcus02's version

**Comment by Gverschueren** [comment]
still have the original problem, can't do anything except put on and off..
I can not install the  

npm install -g --unsafe-perm tmarcus02/homebridge-magichome

code 128
npm ERR! Command failed: /usr/bin/git clone --depth=1 -q -b master git://github.com/tmarcus02/homebridge-magichome.git /root/.npm/_cacache/tmp/git-clone-18e26f1c

can anyone help me?
I'm kinda new to all this :)

**Comment by GatoPharaoh** [comment]
@Gverschueren possible you're just missing sudo?

**Comment by Sunrise17** [comment]
im not missing sudo but still I'm receiving the same fault.

**Comment by GatoPharaoh** [comment]
try updating to the latest npm version
`sudo npm install -g npm`

**Comment by djdom111** [comment]
Hi, thank you for the plugin. I still do have a problem with the white led on the rgbw strip. My controller is 5ch one  (http://amzn.to/2eAljEV). I tried both steve228uk's and tmarcus02's versions with the settings RGBWW and RGBWWv1 respectively. I can turn on and off, change the color of the RGB leds, but the white one is all the time on, and I can't change the brightness. I tried all the setting install/uninstall and all the time is the same. Can you help me, please. Thanks a lot.

**Comment by Zacknetic** [comment]
Yes I can help you. Unfortunately I'm on on vacation now so I can't give you the exact instructions. If you look at my branch of the plugin you will find a edited version which will work.

**Comment by djdom111** [comment]
Lethegrin, thank you for your reply. I found your modified plugin. However when I installed your plugin it doesn't turn on. I tried both settings RGBW and RGBWW. It looks like your modified flux_led.py doesn't work with my controller (I come back to the original plugin, I was able to turn it on and than I change only your flux_led.py  and didn't turn on again. I don't want to disturb in your vacation but maybe you can quickly suggest what I can modify in the file, otherwise I will be waiting for you until you finish your vacation:). One more time thank you for your replay and help.

**Comment by Zacknetic** [comment]
Two questions, does home bridge run after using my plugin? You must use chmod for flux_led.py

Secondly, in your config.json are you setting pure white to true?

**Comment by djdom111** [comment]
Yes, Homebridge runs ok after your plugin, there is also a wheel to change the color and everything what should be. When I press the button to on it shows that is on, but the light is off.  I tried with both settings for purewhite (true and false)

**Comment by Zacknetic** [comment]
Just copy in my flux_led.py and index.js and use chmod on flux_led.py. Google chmod if you're unsure if it's use. ALSO open both files in something like notepad++ and make sure "line endings" are set to "unix" not " windows "

How are you copying the files by the way. What system are you using?

**Comment by djdom111** [comment]
I have a system on Raspberry pi. First time I installed your plugin using this command
sudo npm install -g --unsafe-perm lethegrin/homebridge-magichome 
than I had uninstalled it, and installed the steve's one. After it I copied your file using command
sudo cp and I have the same effect (it doesn't on)

I open the files using Atom.

How to check the line endings. Where shoul I copy your files to chmode them?

**Comment by djdom111** [comment]
One more time thank you for help and a great plugin. I found that andysmithfal's plugin works for me. Lethegrin, thank you and have a great vacations.

**Comment by Zacknetic** [comment]
Hey DJDOM, did you figure it out?

**Comment by djdom111** [comment]
Hi Lethegrin, I did use andysmithfal's plugin. This one works for me without any problems. I don't know what is the difference between yours and his but I wasn't able to figure out yours. One more time thank you for your help :)

**Comment by zandegran** [comment]
andysmithfal's [flux_led.py](https://github.com/andysmithfal/homebridge-magichome/blob/master/flux_led.py) solved the problem. Thanks @djdom111 for finding a solution. The difference is that this does not support prorocol LEDENET as supported by [https://home-assistant.io/components/light.flux_led/](https://home-assistant.io/components/light.flux_led/)
